### ISSUE TITLE
fix(rocm): HSA profiler keeps a CPU core busy while Radiance is idle

### ISSUE THREAD

**Issue by usaloCKLEnTO** [magiccodingman/vllm-radiance#6]
## Problem and tested workaround

After a tiny completed GPU workload, the unmodified published Radiance image leaves one native CPU thread continuously busy at idle on an AMD Radeon AI PRO R9700 (gfx1201). In our two-worker inference deployment this keeps two CPU cores busy, adding approximately 31.5 W of measured physical CPU-package power.

Setting `HSA_TOOLS_DISABLE_REGISTER=1` **before process startup** eliminates the idle spin in our tests. This disables HSA tool/profiler registration; ordinary vLLM metrics and GPU monitoring continue to work. Tools that depend on these profiling hooks would require a separate diagnostic run with registration enabled.

Please consider documenting this workaround for the affected image and investigating the bundled ROCm profiler/runtime interaction. Related upstream symptom report: https://github.com/ROCm/legacy-rocm-build/issues/6522 . We have not established that every case there has this same trigger.

## Affected environment

- Published upstream image: `docker.io/magiccodingman/vllm-radiance@sha256:f728d19a83edd8a019bc6e85f80e8d48099f19e96b87efaddebf8cc22fcf04b1`
- GPU: AMD Radeon AI PRO R9700, gfx1201; Fedora 43 guest with GPU passthrough under Proxmox.
- PyTorch: `2.12.0+rocm7.14`; HIP: `7.14.60850`.
- Loaded runtime: `/opt/rocm/core-7.14/lib/libhsa-runtime64.so.1.21.0`.
- Image manifest rocm-systems source pin: `2b22ab0195cc1461cd9abf3b969e9dd7c10af350`.

This reproduces with **one visible GPU, no model, no vLLM server, and no distributed/RCCL initialization**. We additionally verified it in the unmodified upstream image, independently of our derived image's RCCL changes.

## Reproduction

Save the following as `idle-runtime-probe.py`. It checks GPU arithmetic, synchronizes, then measures per-thread CPU over ten idle seconds. It exits 1 if any thread exceeds 5% CPU.

```python
"""Run inside a ROCm image; fail if a synchronized GPU workload spins at idle."""

import datetime
import json
import os
import time
from pathlib import Path

import torch


def snapshot():
    result = {}
    for path in Path('/proc/self/task').glob('*/stat'):
        fields = path.read_text().rsplit(')', 1)[1].split()
        result[path.parent.name] = (int(fields[11]), int(fields[12]))
    return result


tensor = torch.zeros((256, 256), device='cuda')
assert tensor.sum().item() == 0
# Exercise completed work again before observing idle; avoid declaring a fix
# based solely on initialization with no queue or kernel activity.
assert (tensor + 1).sum().item() == 65536
torch.cuda.synchronize()
before = snapshot()
started = time.monotonic()
time.sleep(10)
elapsed = time.monotonic() - started
after = snapshot()
hz = os.sysconf('SC_CLK_TCK')
threads = [
    {
        'tid': tid,
        'user_pct': 100 * (ticks[0] - before[tid][0]) / hz / elapsed,
        'system_pct': 100 * (ticks[1] - before[tid][1]) / hz / elapsed,
    }
    for tid, ticks in after.items()
    if tid in before
]
passed = max(t['user_pct'] + t['system_pct'] for t in threads) < 5
print(
    json.dumps(
        {
            'at': datetime.datetime.now(datetime.UTC).isoformat(),
            'torch': torch.__version__,
            'hip': torch.version.hip,
            'hsa_tools_disable_register': os.environ.get('HSA_TOOLS_DISABLE_REGISTER'),
            'elapsed_seconds': elapsed,
            'threads': threads,
            'passed': passed,
            'libraries': sorted(
                {
                    line.split()[-1]
                    for line in Path('/proc/self/maps').read_text().splitlines()
                    if 'libhsa-runtime' in line or 'libamdhip64' in line
                }
            ),
        }
    ),
    flush=True,
)
raise SystemExit(0 if passed else 1)
```

Run on a Linux host with rootless Podman GPU access:

```bash
podman run --rm -i \
  --device /dev/kfd --device /dev/dri --group-add keep-groups \
  --security-opt label=disable -e HIP_VISIBLE_DEVICES=0 \
  --entrypoint /opt/vllm/bin/python3 \
  docker.io/magiccodingman/vllm-radiance@sha256:f728d19a83edd8a019bc6e85f80e8d48099f19e96b87efaddebf8cc22fcf04b1 \
  - < idle-runtime-probe.py
```

For the workaround, add `-e HSA_TOOLS_DISABLE_REGISTER=1` before `--entrypoint`.

## Measurements

- Unmodified upstream image, default environment: hottest thread **99.5% CPU**, arithmetic succeeds.
- Our Radiance-derived image: default **99.7%**, registration disabled **0.0%**, default repeated **99.5%**; arithmetic succeeds in all cases.
- Previous PyTorch 2.10 / ROCm 7.2 image: **0.0%** in the same probe. This is a comparison of stacks, not a bisected introducing commit.
- Full two-worker engine, after the same sustained prefill workload: worker idle CPU **100.08% / 100.12% → 0.67% / 0.60%**.
- Physical CPU-package power (30-second integrated RAPL windows): **99.51 W → 67.97 W**. Combined GPU package remained approximately **38 W**. These are not wall-meter readings.
- Same seeded workload, three measured repetitions after warmup: C1 decode **82.39 → 82.77 tok/s**, C4 aggregate **188.84 → 191.18**, C8 aggregate **242.05 → 239.38**, 64K prefill **3239.24 → 3241.49**. This small sample supports retained performance, not a statistical equivalence guarantee.
- Functional checks, warm restart and ordinary vLLM/GPU monitoring passed with the workaround.

## Native-thread evidence

GDB identifies the spinning thread as `rocr::core::Runtime::AsyncEventsLoop`. Inspecting its registered signals shows ordinary wake/queue signals with event IDs 1/4, but the signal registered for `rocprofiler::hsa::profiler_serializer_ready_signal_handler` has **event ID 0 and mailbox 0**, despite being an `InterruptSignal`.

In the image-pinned source:

- [Profiler signal pool preallocates 4096 signals](https://github.com/ROCm/rocm-systems/blob/2b22ab0195cc1461cd9abf3b969e9dd7c10af350/projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/signal_pool.cpp#L68).
- [InterruptSignal can lack an event when allocation fails](https://github.com/ROCm/rocm-systems/blob/2b22ab0195cc1461cd9abf3b969e9dd7c10af350/projects/rocr-runtime/runtime/hsa-runtime/core/runtime/interrupt_signal.cpp#L50).
- [AsyncEventsLoop falls back to polling for a missing event](https://github.com/ROCm/rocm-systems/blob/2b22ab0195cc1461cd9abf3b969e9dd7c10af350/projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp#L1844).

The missing event and the registration workaround are observed. Event-pool exhaustion is a supported hypothesis; we did not instrument the allocation failure itself. Plain `hsa_init()` or `torch.cuda.init()` alone was quiet; creating GPU work was required. `HSA_ENABLE_INTERRUPT=1`, `GPU_MAX_HW_QUEUES=1`, and `HSA_WAIT_ANY_DEBUG=1` did not resolve the spin in our tested configurations.

**Comment by canthonyscott** [comment]
Thank you so much! I had this exact problem. This also fixed it for me. My PC is now silent again!
